### PR TITLE
Adds `Document` proc macro for easy index configuration

### LINKED ISSUE #268: Create a macro to ease the configuration of indexes
**Description**
It would be cool if we could somehow create a kind of builder macro but with the [settings of Meilisearch](https://docs.meilisearch.com/reference/api/settings.html#all-settings).
Some settings do not apply here; you can't set your stopwords or synonyms from a document for example.
But anything related to a field should be handled without any issue I think :thinking: 

**Basic example**
```rust
#[derive(Serialize, Deserialize, Document)]
struct Movie {
  #[document(primary_key)]
  movie_id: u64,
  #[document(displayed, searchable)]
  title: String,
  #[document(displayed)]
  description: String,
  #[document(filterable, sortable, displayed)]
  release_date: String,
  #[document(filterable, displayed)]
  genres: Vec<String>,
}
```
I'm not sure `Document` is the best name here.

This should generate something like that;
```rust
// this trait is not generated, it's just hanging somewhere in our crate
pub trait Document {
  fn generate_settings() -> Settings;
  async fn generate_index() -> Result<Index>;
}

impl Document for Movie {
  fn generate_settings() -> Settings {
    Settings::new()
        .with_displayed_attributes(["title", "description", "release_date", "genres"])
        .with_searchable_attributes(["title"])
        .with_filterable_attributes(["release_date", "genres"])
        .with_sortable_attributes(["release_date"])
  }

  async fn generate_index(client: &Client) -> Result<Index> {
    client.create_index("movie", Some("movie_id"))?
        .wait_for_completion(client)?
        .try_make_index(client)
  }
}
```

The macro should check as many errors as possible at compile time.
Like setting two primary keys or two distinct attributes.

I think that would be a great addition to the crate and make it way more accessible / easier to test.

**Other**

I don't think I'll have the time to work on this issue, so if someone else thinks it's a good addition and wants to implement it, please do not hesitate to ask questions when you're not sure about something!

Also, if you don't know anything about rust macros but would like to learn, you can start by following this workshop; https://github.com/dtolnay/proc-macro-workshop

### PR DESCRIPTION
**Pull request by romilpunetha**
# Pull Request

## Related issue
Fixes #268

## What does this PR do?
- It adds a `Document` macro which eases the process of index configuration by allowing setting `primary_key`, `displayed_attributes`, `searchable_attributes`, `sortable_attributes`, `filterable_attributes` and `distinct_attribute` directly from the struct.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by bidoubiwa** [comment]
Hello @romilpunetha,
Thank you very much for contributing to Meilisearch :heart:.
However, I am not available on the weekend, but I will be back on Monday :blush:.

*PS: This message was sent automatically!*

**Comment by bidoubiwa** [comment]
Hey! thanks a lot for you contribution :) I'll be watching this PR when it changes to ready for review

**Comment by romilpunetha** [comment]
@bidoubiwa it's pending on [this](https://github.com/meilisearch/meilisearch-rust/issues/268#issuecomment-1280037442).

**Comment by bidoubiwa** [comment]
Hey @romilpunetha ! 

Thanks for this very great PR :) 🔥 I'm requiring the engine team to look at it to ensure it is correct. They don't have time this week so it will be reviewed next week. 

Meanwhile, to ensure that we are in time for hacktoberfest, I'm adding the `hacktoberfest accepted` tag so that it counts in your number of PR's!

**Comment by bidoubiwa** [comment]
_This message is sent automatically_

Thanks again for contributing to Meilisearch :heart:
If you are participating in [Hacktoberfest](https://blog.meilisearch.com/contribute-hacktoberfest-2022/), and you would like to receive some gift from Meilisearch too, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

File: meilisearch-index-setting-macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    attribute_key: &str,
+) -> proc_macro2::TokenStream {
+    let ident = syn::Ident::new(attribute_key, proc_macro2::Span::call_site());
+
+    if !attribute.is_empty() {
+        quote! {
+            .#ident(#attribute)
+        }
+    } else {
+        proc_macro2::TokenStream::new()
```
Comment (bidoubiwa) [inline meilisearch-index-setting-macro/src/lib.rs:213]: With the same idea as for vec attributes, we should be able to provide `None`. Unfortunately this is not possible since `with_distinct_attribute` does not accept a `None` value.

I'll open an issue about it

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Could you add some additional tests on some corner cases?

File: meilisearch-index-setting-macro/src/lib.rs (hunk 1 above)
Comment (bidoubiwa) [reply meilisearch-index-setting-macro/src/lib.rs:213]: https://github.com/meilisearch/meilisearch-rust/issues/368

File: meilisearch-index-setting-macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    let fields: &syn::Fields = match ast.data {
+        syn::Data::Struct(ref data) => &data.fields,
+        _ => {
+            return proc_macro::TokenStream::from(
+                syn::Error::new(ast.ident.span(), "Applicable only to struct").to_compile_error(),
+            );
+        }
+    };
+
+    let struct_name = &ast.ident;
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs]: Done.

File: meilisearch-index-setting-macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    proc_macro::TokenStream::from(quote! {
+        #document_implementation
+    })
+}
+
+fn get_document_implementation(
+    struct_ident: &syn::Ident,
+    fields: &syn::Fields,
+) -> proc_macro2::TokenStream {
+    let mut attribute_count: std::collections::HashMap<String, i32> =
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs]: Agreed. I did not think this through. Hashset simplified the implementation.

File: meilisearch-index-setting-macro/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                    }
+                }
+            }
+        }
+    }
+    std::result::Result::Ok(value)
+}
+
+fn get_settings_token_for_list(
+    attributes: &Vec<String>,
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs]: Done

File: meilisearch-index-setting-macro/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    attribute_key: &str,
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs]: Done

File: meilisearch-index-setting-macro/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+) -> proc_macro2::TokenStream {
+    let string_attributes = attributes.iter().map(|attr| {
+        quote! {
+            #attr
+        }
+    });
+
+    let ident = syn::Ident::new(attribute_key, proc_macro2::Span::call_site());
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs]: Done

**Comment by romilpunetha** [comment]
I've updated the implementation to push unfavorable constructs to compiler errors:

1. Handle duplicates on a field.
2. Only commas are allowed as a separator.

File: src/documents.rs [hunk 7]
```diff
@@ -238,17 +248,42 @@ impl<'a> DocumentsQuery<'a> {
...
-    use meilisearch_test_macro::meilisearch_test;
     use serde::{Deserialize, Serialize};
 
+    use meilisearch_index_setting_macro::Document;
+    use meilisearch_test_macro::meilisearch_test;
+
+    use crate::{client::*, indexes::*};
+
+    use super::*;
+
```
Comment (loiclec) [inline src/documents.rs:259]:
```suggestion
use crate as meilisearch_sdk;
```

File: src/documents.rs [hunk 8]
```diff
@@ -1,6 +1,16 @@
...
 
+use crate::settings::Settings;
+use crate::tasks::Task;
+use crate::{errors::Error, indexes::Index};
+
+#[async_trait]
+pub trait Document {
+    fn generate_settings() -> Settings;
+    async fn generate_index(client: &crate::client::Client) -> Result<Index, Task>;
+}
```
Comment (loiclec) [inline src/documents.rs:12]: Since this trait can be derived using a derive proc macro, we should also re-export and document the macro here. You can add the following snippet above the trait declaration, and maybe improve the documentation I came up with if you want :)
```rust

/// Derive the [`Document`](crate::documents::Document) trait.
///
/// ## Field attribute
/// Use the `#[document(..)]` field attribute to generate the correct settings
/// for each field. The available parameters are:
/// - `primary_key` (can only be used once)
/// - `distinct` (can only be used once)
/// - `searchable`
/// - `displayed`
/// - `filterable`
/// - `sortable`
///
/// ## Index name
/// The name of the index will be the name of the struct converted to snake case.
///
/// ## Sample usage:
/// ```
/// use serde::{Serialize, Deserialize};
/// use meilisearch_sdk::documents::Document;
/// use meilisearch_sdk::settings::Settings;
/// use meilisearch_sdk::indexes::Index;
/// use meilisearch_sdk::client::Client;
///
/// #[derive(Serialize, Deserialize, Document)]
/// struct Movie {
///     #[document(primary_key)]
///     movie_id: u64,
///     #[document(displayed, searchable)]
///     title: String,
///     #[document(displayed)]
///     description: String,
///     #[document(filterable, sortable, displayed)]
///     release_date: String,
///     #[document(filterable, displayed)]
///     genres: Vec<String>,
/// }
///
/// async fn usage(client: Client) {
///     // Default settings with the distinct, searchable, displayed, filterable, and sortable fields set correctly.
///     let settings: Settings = Movie::generate_settings();
///     // Index created with the name `movie` and the primary key set to `movie_id`
///     let index: Index = Movie::generate_index(&client).await.unwrap();
/// }
/// ```
pub use meilisearch_index_setting_macro::Document;
```

File: src/documents.rs [hunk 9]
```diff
@@ -238,17 +248,42 @@ impl<'a> DocumentsQuery<'a> {
...
+        title: String,
+        #[document(displayed)]
+        description: String,
+        #[document(filterable, sortable, displayed)]
+        release_date: String,
+        #[document(filterable, displayed)]
+        genres: Vec<String>,
+    }
+
+    #[derive(Document)]
```
Comment (loiclec) [inline src/documents.rs:282]:
```suggestion
    #[allow(unused)]
    #[derive(Document)]
```

File: src/documents.rs [hunk 10]
```diff
@@ -238,17 +248,42 @@ impl<'a> DocumentsQuery<'a> {
...
+
+    use super::*;
+
     #[derive(Debug, Serialize, Deserialize, PartialEq)]
     struct MyObject {
         id: Option<usize>,
         kind: String,
     }
 
+    #[derive(Document)]
```
Comment (loiclec) [inline src/documents.rs:266]:
```suggestion
    #[allow(unused)]
    #[derive(Document)]
```

**Review by loiclec — CHANGES_REQUESTED** [review]
Hi @romilpunetha ,

I am sorry for the long wait before the review, and thanks a lot for starting this PR. It looks very good :) At the moment, due to a small oversight, it is not possible to use the proc macro  in other crates. I left a few suggestions to resolve these problems and then I think it should be good to go 😄

File: meilisearch-index-setting-macro/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    fields: &syn::Fields,
+) -> proc_macro2::TokenStream {
+    let mut attribute_count: std::collections::HashMap<String, i32> =
+        std::collections::HashMap::new();
+    let mut primary_key_attribute: String = "".to_string();
+    let mut distinct_key_attribute: String = "".to_string();
+    let mut displayed_attributes: Vec<String> = vec![];
+    let mut searchable_attributes: Vec<String> = vec![];
+    let mut filterable_attributes: Vec<String> = vec![];
+    let mut sortable_attributes: Vec<String> = vec![];
```
[the remark this answers is not in the document]
Comment (romilpunetha) [reply meilisearch-index-setting-macro/src/lib.rs:40]: Have handled this as a compiler error.

File: src/documents.rs (hunk 8 above)
Comment (romilpunetha) [reply src/documents.rs:12]: This is throwing an error:

```
error[E0433]: failed to resolve: unresolved import
  --> src/documents.rs:27:34
   |
10 | #[derive(Serialize, Deserialize, Document)]
   |                                  ^^^^^^^^
   |                                  |
   |                                  unresolved import
   |                                  help: a similar path exists: `meilisearch_sdk::client`
   |
   = note: this error originates in the derive macro `Document` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by bidoubiwa** [comment]
Hey @romilpunetha ! Thanks a lot for this contribution :) We are going to merge it in a temp branch so that @loiclec can make the last changes. After which it will be merged on main.

🔥🔥🔥 Thanks for this great contribution

**Review by bidoubiwa — APPROVED** [review]

**Comment by romilpunetha** [comment]
@loiclec @bidoubiwa  Thanks for all the help. I'm new to rust and this PR was a good starting point for me.
